File: backend/src/main/java/com/freshmart/backend/service/interfaces/cart_management_module/CartService.java
package com.freshmart.backend.service.interfaces.cart_management_module;

import java.util.List;

import com.freshmart.backend.dto.request.cart_management_module.AddToCartRequest;
import com.freshmart.backend.dto.request.cart_management_module.UpdateCartItemRequest;
import com.freshmart.backend.dto.response.cart_management_module.CartItemResponse;

/** Implements UC11-UC14 (Add to Cart, View Cart, Update Quantity, Remove Item). */
public interface CartService {

    /** Adds quantity to an existing row for the same product, or creates a new one. */
    CartItemResponse addToCart(Long userId, AddToCartRequest request);

    List<CartItemResponse> viewCart(Long userId);

    /**
     * Per UC13's alternative flow: decreasing quantity to 0 forwards to
     * Remove Item instead of updating. Returns {@code null} when the item
     * was removed this way.
     */
    CartItemResponse updateQuantity(Long userId, Long cartItemId, UpdateCartItemRequest request);

    boolean removeItem(Long userId, Long cartItemId);
}

File: frontend/src/features/cart_management_module/services/cartService.js
import apiClient from "../../../common/api";

/**
 * Calls POST /api/v1/customer/cart. Returns the CartItemResponse payload
 * unwrapped from the ApiResponse envelope.
 */
export async function addToCart({ productId, quantity }) {
  const { data } = await apiClient.post("/customer/cart", { productId, quantity });
  return data.data;
}

/** Calls GET /api/v1/customer/cart. Returns the list of CartItemResponse. */
export async function viewCart() {
  const { data } = await apiClient.get("/customer/cart");
  return data.data;
}

/**
 * Calls PUT /api/v1/customer/cart/{cartItemId}. Per UC13's alternative
 * flow, quantity 0 forwards to Remove Item on the backend — data.data is
 * null in that case.
 */
export async function updateCartItemQuantity(cartItemId, quantity) {
  const { data } = await apiClient.put(`/customer/cart/${cartItemId}`, { quantity });
  return data.data;
}

/** Calls DELETE /api/v1/customer/cart/{cartItemId}. */
export async function removeCartItem(cartItemId) {
  const { data } = await apiClient.delete(`/customer/cart/${cartItemId}`);
  return data.data;
}
